generative_models/VAEGAN-PYTORCH/main.py

```diff
@@ -0,0 +1,120 @@
+import torch
+import torch.nn as nn
+import torchvision
+import torchvision.transforms as transforms
+import numpy as np
+import matplotlib.pyplot as plt
+from torch.autograd import Variable
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+torch.autograd.set_detect_anomaly(True)
+from dataloader import dataloader
+from models import VAE_GAN,Discriminator
+from utils import show_and_save,plot_loss
+
+data_loader=dataloader(64)
+gen=VAE_GAN().to(device)
+discrim=Discriminator().to(device)
+real_batch = next(iter(data_loader))
+show_and_save("training" ,make_grid((real_batch[0]*0.5+0.5).cpu(),8))
+
+epochs=25
+lr=3e-4
+alpha=0.1
+gamma=15
+
+criterion=nn.BCELoss().to(device)
+optim_E=torch.optim.RMSprop(gen.encoder.parameters(), lr=lr)
+optim_D=torch.optim.RMSprop(gen.decoder.parameters(), lr=lr)
+optim_Dis=torch.optim.RMSprop(discrim.parameters(), lr=lr*alpha)
+z_fixed=Variable(torch.randn((64,128))).to(device)
+x_fixed=Variable(real_batch[0]).to(device)
+
+for epoch in range(epochs):
+  prior_loss_list,gan_loss_list,recon_loss_list=[],[],[]
+  dis_real_list,dis_fake_list,dis_prior_list=[],[],[]
+  for i, (data,_) in enumerate(data_loader, 0):
+    bs=data.size()[0]
+    
+    ones_label=Variable(torch.ones(bs,1)).to(device)
```

Comment:
add one line comment (like training discriminator or generator ) to make it read friendly